source/mainnet/smart-contracts/guides/factory-pattern.rst

```diff
@@ -0,0 +1,392 @@
+and contract name against expected values. Together, the module reference and contract name uniquely
+identify the code of the smart contract instance.
+
+Getting the module reference and contract name is done using the host functions
+``contract_module_reference`` and ``contract_name``, respectively. Note: both of these functions
+are introduced in protocol version 7, and will not work while the chain is running an earlier
+protocol version.
+
+In this example, the factory and product contracts are defined in the same module.
+Thus, to check that the module reference of the product is correct, it is sufficient to check that
+it is the same as the module reference of the factory, as follows:
+
+.. code-block:: Rust
+
+        // We can depend upon getting the module reference for our own contract.
+        let self_module_ref = host.contract_module_reference(ctx.self_address()).unwrap();
+        // Check the product module is the same as our own module.
+        let product_module_ref = host
+            .contract_module_reference(product_address)
+            .or(Err(FactoryError::NonExistentProduct))?;
+        ensure_eq!(self_module_ref, product_module_ref, FactoryError::InvalidProduct);
+
+If the product were instead defined in a different module from the factory, then the expected
+module reference for the product could be hard-coded into the factory. Alternatively, if the
+module rereference for the product is to be determined later, it could be passed as a parameter
+when creating the ``factory`` instance.
+
+If the module defining the product is known to only contain one smart contract, then checking the
+module reference is sufficient for identifying the code of the product smart contract. In this case,
+however, both the factory and product contracts are defined in the same module, so it is necessary
+to also check the contract name. This is achieved as follows:
+
+.. code-block:: Rust
+
+        let product_name =
+            host.contract_name(product_address).or(Err(FactoryError::NonExistentProduct))?;
+        ensure_eq!(product_name, PRODUCT_INIT_NAME, FactoryError::InvalidProduct);
+
+Now the contract is known to be an instance of ``product``, the next step is to call
+``initialize``. In this example, ``initialize`` takes as a parameter that is the index assigned to
+it, which will be the current value of ``next_product`` in the state.
+
+.. code-block:: Rust
+
+        let next_product = host.state().next_product;
+        host.invoke_contract(
+            &product_address,
+            &next_product,
+            EntrypointName::new_unchecked(PRODUCT_INITIALIZE_ENTRYPOINT),
+            Amount::zero(),
+        )
+        .or(Err(FactoryError::InitializeFailed))?;
+
+Here, it is assumed that ``initialize`` will fail, for instance, if it is called on a product that
+has previously been initialized.
+It only remains to update the factory contract's state:
+
+.. code-block:: Rust
+
+        let state = host.state_mut();
```

Comment:
The usual smart contract best practice pattern is to do a `check-effect-interact` pattern. This means we should do the state update in the factory before calling an external contract. We can consider to change the order in this function in the factory example.